File: Intermediate/23.Implement a min stack that supports push, pop, and retrieving the minimum element in constant time/Solution.java
import java.util.Scanner;
import java.util.Stack;

class MinStack {
    private Stack<Integer> mainStack = new Stack<>();
    private Stack<Integer> minStack = new Stack<>();

    public void push(int value) {
        mainStack.push(value);
        if (minStack.isEmpty() || value <= minStack.peek()) {
            minStack.push(value);
        }
        System.out.println("Pushed: " + value);
    }

    public void pop() {
        if (mainStack.isEmpty()) {
            System.out.println("Stack underflow.");
            return;
        }
        if (mainStack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        System.out.println("Popped: " + mainStack.pop());
    }

    public int getMin() {
        if (minStack.isEmpty()) {
            System.out.println("Stack is empty.");
            return Integer.MAX_VALUE;
        }
        return minStack.peek();
    }
}

public class MinStackDemo {
    public static void main(String[] args) {
        MinStack ms = new MinStack();
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n1. Push\n2. Pop\n3. Get Minimum\n4. Exit\nEnter your choice: ");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter value to push: ");
                    int value = sc.nextInt();
                    ms.push(value);
                    break;
                case 2:
                    ms.pop();
                    break;
                case 3:
                    System.out.println("Minimum: " + ms.getMin());
                    break;
                case 4:
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}

File: Intermediate/23.Implement a min stack that supports push, pop, and retrieving the minimum element in constant time/Solution.py
class MinStack:
    def __init__(self):
        self.main_stack = []
        self.min_stack = []

    def push(self, value):
        self.main_stack.append(value)
        if not self.min_stack or value <= self.min_stack[-1]:
            self.min_stack.append(value)
        print(f"Pushed: {value}")

    def pop(self):
        if not self.main_stack:
            print("Stack underflow.")
            return
        if self.main_stack[-1] == self.min_stack[-1]:
            self.min_stack.pop()
        print(f"Popped: {self.main_stack.pop()}")

    def get_min(self):
        if not self.min_stack:
            print("Stack is empty.")
            return float('inf')
        return self.min_stack[-1]


if __name__ == "__main__":
    ms = MinStack()
    while True:
        print("\n1. Push\n2. Pop\n3. Get Minimum\n4. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter value to push: "))
            ms.push(value)
        elif choice == 2:
            ms.pop()
        elif choice == 3:
            print(f"Minimum: {ms.get_min()}")
        elif choice == 4:
            break
        else:
            print("Invalid choice.")
